File: deep_search-dev/svs_deepsearch.py
import uvicorn
import traceback
import time
import asyncio
from art import text2art
from fastapi import FastAPI, HTTPException, Request
from pathlib import Path
from logs.logger import define_log_level
from pydantic import BaseModel
from deepsearch.main_deepsearch import DeepSearch
from deepsearch.utils.parse_data import verify_params


app = FastAPI()
VERSION = '0.1'
project_root = Path(__file__).resolve().parent


class DeepSearchRequest(BaseModel):
    topic: str = ""
    

@app.post("/api/deep_search")
async def deep_search(request: DeepSearchRequest, fastapi_request: Request):
    """
    处理 deepsearch 请求的主接口

    Args:
        request: FastAPI请求对象

    Returns:
        JSON响应
    """
    # 参数
    topic = request.topic

    # 创建日志
    logger, log_name = define_log_level(project_root, topic)
    logger.info(f"日志已创建：{log_name}")

    try:
        start = time.time()

        # 参数校验
        topic = verify_params(topic)
        logger.info(f"参数校验成功！主题：{topic}")

        # 创建带取消功能的任务
        deepsearch_task = asyncio.create_task(
            DeepSearch(topic, project_root, logger).run()
        )

        # 添加请求取消监听
        stop_event = asyncio.Event()
        request_task = asyncio.create_task(request_disconnect_handler(fastapi_request, stop_event))

        # 等待任务完成或取消
        done, pending = await asyncio.wait(
            [deepsearch_task, request_task],
            return_when=asyncio.FIRST_COMPLETED
        )

        # 处理取消信号
        if not deepsearch_task.done():
            logger.warning("用户取消了请求，正在终止任务...")
            deepsearch_task.cancel()
            try:
                await deepsearch_task
            except asyncio.CancelledError:
                logger.info("任务已成功终止")
            return {
                "status": 499,  # Nginx定义的客户端关闭连接状态码
                "msg": "请求已被用户取消"
            }

        # 获取正常结果
        result = deepsearch_task.result()

        end = time.time()
        logger.info(f"深度搜索总共耗时：{end - start}s")

        return {
            "status": 200,
            "time": end - start,
            "msg": "success",
            "data": result
        }
    except HTTPException as e:
        logger.error(e.detail)
        return {
            "status": 400,
            "msg": 'fail',
            "data": e.detail
        }
    except Exception as e:
        logger.error(traceback.format_exc())
        return {
            "status": 500,
            "msg": "fail",
            "data": str(e)
        }


async def request_disconnect_handler(request: Request, stop_event: asyncio.Event):
    """监听客户端连接状态"""
    while not stop_event.is_set():
        if await request.is_disconnected():
            stop_event.set()
            return
        await asyncio.sleep(0.1)


if __name__ == "__main__":
    print(text2art('DEEPSEARCH'))
    print('Version:', VERSION)

    uvicorn.run(app, host="0.0.0.0", port=7396, reload=False)

File: Agent/deep_search-dev-go/svs_deepsearch.go
package main

import (
	"context"
	"deepsearch/api"
	"deepsearch/deepsearch"
	"deepsearch/logs"
	"deepsearch/deepsearch/utils"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"time"

	"github.com/joho/godotenv"
)

// 版本号
const VERSION = "0.1"

// DeepSearchRequest 定义请求数据结构
type DeepSearchRequest struct {
	Topic string `json:"topic"`
}

// DeepSearchResponse 定义响应数据结构
type DeepSearchResponse struct {
	Status int         `json:"status"`
	Time   float64     `json:"time,omitempty"`
	Msg    string      `json:"msg"`
	Data   interface{} `json:"data,omitempty"`
}

// 全局变量，项目根目录
var projectRoot string

// 初始化函数
func init() {
	// 获取当前执行文件所在目录的绝对路径
	var err error
	projectRoot, err = filepath.Abs(filepath.Dir(os.Args[0]))
	if err != nil {
		panic(err)
	}
}

func main() {
	// 加载环境变量
	err := godotenv.Load()
	if err != nil {
		log.Println("警告: 加载.env文件失败，将使用系统环境变量")
	}

	// 打印ASCII艺术字体的DEEPSEARCH
	printAsciiArt("DEEPSEARCH")
	fmt.Printf("Version: %s\n", VERSION)

	// 设置HTTP路由
	http.HandleFunc("/api/deep_search", deepSearchHandler)

	// 启动HTTP服务器
	fmt.Println("Starting server on :7396")
	err = http.ListenAndServe(":7396", nil)
	if err != nil {
		fmt.Printf("Server failed to start: %v\n", err)
	}
}

// deepSearchHandler 处理深度搜索请求
func deepSearchHandler(w http.ResponseWriter, r *http.Request) {
	// 只处理POST请求
	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	// 解析请求体
	var req DeepSearchRequest
	err := json.NewDecoder(r.Body).Decode(&req)
	if err != nil {
		sendErrorResponse(w, 400, "Invalid request format", err.Error())
		return
	}

	// 获取topic参数
	topic := req.Topic

	// 创建日志 - 在接收到请求时才创建日志，与Python版本一致
	logger, logName := logs.DefineLogLevel(projectRoot, topic, "info", "info")
	logger.Infof("日志已创建：%s", logName)

	// 初始化全局LLM实例
	api.InitLLM(logger)

	// 创建上下文，用于处理请求取消
	ctx, cancel := context.WithCancel(r.Context())
	defer cancel()

	// 使用通道来处理结果和错误
	resultChan := make(chan interface{})
	errorChan := make(chan error)

	// 启动goroutine处理深度搜索
	go func() {
		defer func() {
			if r := recover(); r != nil {
				errorChan <- fmt.Errorf("panic: %v", r)
			}
		}()

		start := time.Now()

		// 参数校验
		topic, err := utils.VerifyParams(topic)
		if err != nil {
			errorChan <- err
			return
		}
		logger.Infof("参数校验成功！主题：%s", topic)

		// 创建深度搜索实例并运行
		ds := deepsearch.NewDeepSearch(projectRoot, logger)
		result, err := ds.Run(ctx, topic, "deepsearch")
		if err != nil {
			errorChan <- err
			return
		}

		// 计算耗时
		elapsed := time.Since(start).Seconds()
		logger.Infof("深度搜索总共耗时：%fs", elapsed)

		// 返回结果
		response := DeepSearchResponse{
			Status: 200,
			Time:   elapsed,
			Msg:    "success",
			Data:   result,
		}
		resultChan <- response
	}()

	// 启动goroutine监听请求取消
	go func() {
		select {
		case <-ctx.Done():
			// 请求被取消
			logger.Warning("用户取消了请求，正在终止任务...")
			errorChan <- fmt.Errorf("request cancelled")
			return
		case <-r.Context().Done():
			// 客户端断开连接
			logger.Warning("客户端断开连接")
			errorChan <- fmt.Errorf("request cancelled")
			return
		}
	}()

	// 处理结果或错误
	select {
	case result := <-resultChan:
		sendJSONResponse(w, result)
	case err := <-errorChan:
		if err.Error() == "request cancelled" {
			sendErrorResponse(w, 499, "请求已被用户取消", nil)
		} else {
			sendErrorResponse(w, 500, "fail", err.Error())
		}
	}
}

// sendJSONResponse 发送JSON响应
func sendJSONResponse(w http.ResponseWriter, data interface{}) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(data)
}

// sendErrorResponse 发送错误响应
func sendErrorResponse(w http.ResponseWriter, status int, msg string, data interface{}) {
	response := DeepSearchResponse{
		Status: status,
		Msg:    msg,
		Data:   data,
	}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(response)
}

// printAsciiArt 打印ASCII艺术字体
func printAsciiArt(text string) {
	// 简单的ASCII艺术字体
	art := `
    ____            __  __                       __      
   / __ \___  _____/ /_/ /_  ____  ____     ____/ /______
  / /_/ / _ \/ ___/ __/ __ \/ __ \/ __ \   / __  / ___/
 / ____/  __(__  ) /_/ / / / /_/ / / / /  / /_/ (__  ) 
/_/    \___/____/\__/_/ /_/\____/_/ /_/   \__,_/____/  
`
	fmt.Println(art)
}
